tests/xdc/history_replication_conflict_test.go

```diff
@@ -194,6 +199,68 @@ func (s *historyReplicationConflictTestSuite) TestConflictResolutionReappliesSig
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, testTimeout)
+	defer cancel()
+	sdkClient1, sdkClient2 := s.createSdkClients()
+
+	runId := s.startWorkflowAndEnterSplitBrainState(ctx, sdkClient1)
+
+	// Both clusters now believe they are active and hence both will accept an update.
+
+	// Send updates
+	s.sendUpdateAndProcessWFT(ctx, runId, "cluster1-update", "cluster1-update-input", sdkClient1, s.cluster1)
+	s.sendUpdateAndProcessWFT(ctx, runId, "cluster2-update", "cluster2-update-input", sdkClient2, s.cluster2)
+
+	// cluster1 has accepted an update
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowTaskStarted
+	4 WorkflowTaskCompleted
+	5 WorkflowExecutionUpdateAccepted {"AcceptedRequest": {"Input": {"Args": {"Payloads": [{"Data": "\"cluster1-update-input\""}]}}}}
+	`, s.getHistory(ctx, s.cluster1, runId), []int{1, 1, 1, 1, 1})
+
+	// cluster2 has also accepted an update (events have failover version 2 since they are endogenous to cluster 2)
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowTaskStarted
+	4 WorkflowTaskCompleted
+	5 WorkflowExecutionUpdateAccepted {"AcceptedRequest": {"Input": {"Args": {"Payloads": [{"Data": "\"cluster2-update-input\""}]}}}}
+	`, s.getHistory(ctx, s.cluster2, runId), []int{1, 1, 2, 2, 2})
+
+	// Execute pending history replication tasks. Both clusters believe they are active, therefore each cluster sends
+	// its update to the other, triggering conflict resolution.
+	s.executeHistoryReplicationTasksUntil(ctx, "cluster1", enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ACCEPTED)
+	s.executeHistoryReplicationTasksUntil(ctx, "cluster2", enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ACCEPTED)
+
+	// cluster1 has received a update with failover version 2 which superseded its own update.
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowTaskStarted
+	4 WorkflowTaskCompleted
+	5 WorkflowExecutionUpdateAccepted {"AcceptedRequest": {"Input": {"Args": {"Payloads": [{"Data": "\"cluster2-update-input\""}]}}}}
+	`, s.getHistory(ctx, s.cluster1, runId), []int{1, 1, 2, 2, 2})
+
+	// cluster2 has reapplied the accepted update from cluster 1 on top of its own update, changing it from state
+	// Accepted to state Requested, since it must be submitted to the validator on the new branch.
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowTaskStarted
+	4 WorkflowTaskCompleted
+	5 WorkflowExecutionUpdateAccepted {"AcceptedRequest": {"Input": {"Args": {"Payloads": [{"Data": "\"cluster2-update-input\""}]}}}}
+	6 WorkflowExecutionUpdateRequested {"Request": {"Input": {"Args": {"Payloads": [{"Data": "\"cluster1-update-input\""}]}}}}
+	7 WorkflowTaskScheduled
+	`, s.getHistory(ctx, s.cluster2, runId), []int{1, 1, 2, 2, 2, 2, 2})
```

Comment:
I am working on adding further related test coverage. E.g.

1. The conflicting updates have the same update ID vs different update IDs. Currently same update IDs will be rejected at acceptance but I need to look into how we want to handle this.
2. Continue the test to process the update after reapplying: see https://github.com/temporalio/temporal/commit/2e82093737959eaa5a5ecba929bc099e5000e71a

Reply:
👍  
We also need to check if the update accepted message responded from worker will actually be recorded by the server (i.e. update registry knows about this reapplied update).